### PR TITLE
2.0/multi-source queries

### PR DESCRIPTION
**Pull request by baseTwo**
ℹ️Work for issue #197 and  #220 

# Fixing multi-source queries with cross joins
* Merging `Single`- and `MultiSourceQuery` methods into one `QueryImpl`
* Use a new CQL operator `CrossJoin` instead of repeated nesting of `SelectMany` -- more efficient since it doesn't have to create a new two-item tuple type between each consecutive foreach
* A new `TypeFor` method that can construct cql tuples from a dictionary of (string name, Type type) 
* Allow for up to 8 source on a cross-join on the cql operator context. CMS measures have only up to 7 sources. Will need to expand on this in the future with nested CrossJoins.

# Other Changes
## Dumping debug information on queries to the console. 
Dump the source and cql info to the log for queries. For this it was necessary to know how to get from the library to the cql file, so the library type was modified so that the OriginalFileName is set to the elm file from which it was loaded, and then using a convention to get from the elm file to the cql file
image

## `CqlOperator` Enum:
* List-based operations to start with `List-` for consistency: `ListElementAt`, `ListIndecOf`, `ListSort`
* Added `CrossJoin`

## Extensions
* `SelectToArray` pre-allocates an array to a known size of the source enumeration, then selects the source items into the target array. Less wasteful than calling `enumerable.ToArray`

## `TypeResolver`
* Updated code comments
* Renamed parameter `@throw` to `throwError` (consistency)

## CQL Errors
* Renamed parameters for type names to be consistent: `TypeName`
* Made some parameters nullable

## Dumpify Nuget
* A debug package making it easier to dump objects while debugging to the console. (Similar output than linqpad)

## EditorConfig
Added two config settings for ReSharper, to set max args/params per line to 2, after which they will be put on multilines

### CODE REVIEW

**Comment by ewoutkramer** [comment]
Also fixed #197.

File: Cql/Cql.Compiler/ExpressionBuilder.Query.cs [hunk 1]
```diff
@@ -1,57 +1,66 @@
...
         }
 
-        var @return = source;
+        ParameterExpression scopeParameter;
 
-        Type elementType = _typeManager.Resolver.GetListElementType(@return.Type, @throw: true)!;
+        var (@return, isSourcesPromoted) = ctx.ProcessQuerySources(query);
+        var isAnySourcePromoted = isSourcesPromoted.Any(isSourcePromoted => isSourcePromoted);
+        var returnType = ctx._typeManager.Resolver.GetListElementType(@return.Type, true)!;
+        var returnElementType = ctx._typeManager.Resolver.GetListElementType(@return.Type, true)!;
```
Comment (baseTwo) [inline Cql/Cql.Compiler/ExpressionBuilder.Query.cs:33]: Exactly the same type, remote it and the check below

File: Cql/Cql.Compiler/ExpressionBuilder.Query.cs [hunk 2]
```diff
@@ -86,339 +95,370 @@ protected Expression SingleSourceQuery(Query query)
...
+        //    IEnumerable<A> a = ...;
+        //    IEnumerable<B> b = ...;
+        //    IEnumerable<c> c = ...;
+        //    IEnumerable<(A, B, C)> crossJoinedValueTupleResults = CrossJoin<A, B, C>(a, b, c);
+
+        var tryPromoteSourceExpressions = sourceExpressions.SelectToArray(expr => TryPromoteSource(expr));
+        var promotedSourceExpressions = tryPromoteSourceExpressions.SelectToArray(s => s.source);
+        var isSourcesPromoted = tryPromoteSourceExpressions.SelectToArray(s => s.isSourcePromoted);
+
+        if (promotedSourceExpressions.Any(promotedSource => !IsOrImplementsIEnumerableOfT(promotedSource.Type)))
```
Comment (ewoutkramer) [inline Cql/Cql.Compiler/ExpressionBuilder.Query.cs]: Based on the lines above, how can any of the sources still not be promoted?

**Review by ewoutkramer — CHANGES_REQUESTED** [review]

File: Cql/Cql.Compiler/ExpressionBuilder.Query.cs (hunk 2 above)
Comment (baseTwo) [reply Cql/Cql.Compiler/ExpressionBuilder.Query.cs]: I changed the wording here to indicate if the original sources were singletons.

**Review by ewoutkramer — APPROVED** [review]

**Comment by baseTwo** [comment]
Replaced with another PR #285

**Comment by baseTwo** [comment]
Replaced with another PR #285

**Pull request closed without merging**